packages/web-frontend/src/selectors/measureSelectors.js

```diff
@@ -34,139 +32,143 @@ const displayInfoCache = createCachedSelector(
+          measureData[measure.organisationUnitCode] = {
+            ...measureData[measure.organisationUnitCode],
+            ...measure,
+          };
+        });
+      }
+    });
+    return Object.keys(measureData).length === 0 ? undefined : Object.values(measureData);
   },
 );
 
-const selectDisplayLevelAncestor = createSelector(
-  [
-    state => selectCountryHierarchy(state, selectCurrentOrgUnitCode(state)),
-    selectCurrentOrgUnitCode,
-    selectMeasureOptionsByDisplayedMapOverlays,
-  ],
-  (country, currentOrganisationUnitCode, measureOptions) => {
-    if (!country || !currentOrganisationUnitCode || !measureOptions) {
-      return undefined;
-    }
+const selectMeasureLevel = createSelector(
+  [state => state.map.measureInfo, (_, mapOverlayIds) => mapOverlayIds],
+  (measureInfo, mapOverlayIds) => {
+    let measureLevelArray = [];
 
-    const displayOnLevel = measureOptions.map(option => option.displayOnLevel).find(level => level);
-    if (!displayOnLevel) {
-      return undefined;
-    }
+    mapOverlayIds.forEach(mapOverlayId => {
+      const { measureLevel: selectedMeasureLevel = [] } = measureInfo[mapOverlayId] || {};
+      measureLevelArray = [...measureLevelArray, ...selectedMeasureLevel];
+    });
 
-    return selectAncestors(country, currentOrganisationUnitCode, displayOnLevel).find(
-      ancestor => ancestor.type === displayOnLevel,
-    );
+    return [...new Set(measureLevelArray)];
   },
 );
 
 export const selectHasPolygonMeasure = createSelector(
-  [state => state.map.measureInfo],
-  (measureInfo = {}) => {
-    return (
-      measureInfo.measureOptions &&
-      measureInfo.measureOptions.some(option => POLYGON_MEASURE_TYPES.includes(option.type))
-    );
+  [state => state.map.measureInfo, state => state.map.displayedMapOverlays],
+  (measureInfo = {}, displayedMapOverlays) => {
+    for (const displayedMapOverlayId of displayedMapOverlays) {
+      if (
+        measureInfo[displayedMapOverlayId]?.measureOptions &&
+        measureInfo[displayedMapOverlayId].measureOptions.some(option =>
+          POLYGON_MEASURE_TYPES.includes(option.type),
+        )
+      ) {
+        return true;
+      }
+    }
+    return false;
   },
 );
 
-export const selectAllMeasuresWithDisplayInfo = createSelector(
+export const selectMeasuresWithDisplayInfo = createSelector(
   [
-    state => selectActiveProjectCountries(state),
-    state => selectCurrentProjectCode(state),
-    state => selectCountryHierarchy(state, state.map.measureInfo.currentCountry),
-    state => state.map.measureInfo.measureData,
-    selectMeasureOptionsByDisplayedMapOverlays,
-    state => state.map.measureInfo.hiddenMeasures,
-    state => state.map.measureInfo.currentCountry,
-    state => state.map.measureInfo.measureLevel,
+    selectActiveProjectCountries,
+    selectCurrentProjectCode,
+    state => selectCountryHierarchy(state, state.map.currentCountry),
+    (state, mapOverlayIds) => selectMeasureData(state, mapOverlayIds),
```

Comment:
Btw no need to change it, but this should also work:

```suggestion
    selectMeasureData,
```

All arguments will be passed to the callbacks by JS